expression/builtin_compare.go

```diff
@@ -284,98 +291,404 @@ func (b *builtinCoalesceDurationSig) evalDuration(row []types.Datum) (res types.
 	return res, isNull, errors.Trace(err)
 }
 
+// temporalWithDateAsNumTypeClass makes DATE, DATETIME, TIMESTAMP pretend to be numbers rather than strings.
+func temporalWithDateAsNumTypeClass(argTp *types.FieldType) (argTypeClass types.TypeClass, isStr bool, isTemporalWithDate bool) {
+	argTypeClass = argTp.ToClass()
```

Comment:
can we use `evalTp` instead of `TypeClass` ?

Reply:
The modification will influence the implementation of `getCmpType`,
thus influence the functions which use it.
Better to change this together.